crates/admin/src/cluster_controller/logs_controller.rs

```diff
@@ -0,0 +1,764 @@
+use restate_types::logs::builder::{ChainBuilder, LogsBuilder};
+use restate_types::logs::metadata::{
+    Chain, LogletConfig, LogletParams, Logs, ProviderKind, SegmentIndex,
+};
+use restate_types::logs::{LogId, Lsn};
+use restate_types::metadata_store::keys::BIFROST_CONFIG_KEY;
+use restate_types::nodes_config::Role;
+use restate_types::partition_table::PartitionTable;
+use restate_types::replicated_loglet::{NodeSet, ReplicatedLogletParams, ReplicationProperty};
+use restate_types::{Version, Versioned};
+use std::collections::BTreeMap;
+use std::num::NonZeroU8;
+use tokio::task::JoinSet;
+use tracing::debug;
+
+#[derive(Debug)]
+enum LogsState {
+    Provisioning {
+        log_id: LogId,
+        provider_kind: ProviderKind,
+    },
+    Running {
+        configuration: Option<LogletConfiguration>,
+        segment_index: SegmentIndex,
+    },
+    Sealing {
+        configuration: Option<LogletConfiguration>,
+        segment_index: SegmentIndex,
+    },
+    Sealed {
+        configuration: LogletConfiguration,
+        segment_index: SegmentIndex,
+        seal_lsn: Lsn,
+    },
+}
+
+impl LogsState {
+    fn try_seal(&mut self, segment_index_to_seal: SegmentIndex, seal_lsn: Lsn) {
+        match self {
+            LogsState::Provisioning { .. } => {}
+            LogsState::Running {
+                configuration,
+                segment_index,
+            } => {
+                if *segment_index == segment_index_to_seal {
+                    *self = LogsState::Sealed {
+                        configuration: configuration
+                            .take()
+                            .expect("requires previous configuration"),
+                        segment_index: segment_index_to_seal,
+                        seal_lsn,
+                    }
+                }
+            }
+            LogsState::Sealing {
+                configuration,
+                segment_index,
+            } => {
+                if *segment_index == segment_index_to_seal {
+                    *self = LogsState::Sealed {
+                        configuration: configuration
+                            .take()
+                            .expect("requires previous configuration"),
+                        segment_index: segment_index_to_seal,
+                        seal_lsn,
+                    }
+                }
+            }
+            LogsState::Sealed { .. } => {}
+        }
+    }
+
+    fn try_sealing(&mut self) {
+        match self {
+            LogsState::Provisioning { .. } => {}
+            LogsState::Running {
+                configuration,
+                segment_index,
+            } => {
+                *self = LogsState::Sealing {
+                    configuration: configuration.take(),
+                    segment_index: *segment_index,
+                }
+            }
+            LogsState::Sealing { .. } => {}
+            LogsState::Sealed { .. } => {}
+        }
+    }
+
+    fn try_provisioning(
```

Comment:
All the `try_*` functions to me imply actuating effects, but we're only just building up configuration at this point.

Reply:
True. I've chosen `try_` to indicate that those methods don't transition all states when being called: "Try" to transition if you are in the right state. Do you have a suggestion how to name them?

Reply:
I might have picked `build_*` for this?